build-support/bin/check_header.py

```diff
@@ -62,21 +71,28 @@ def create_parser() -> argparse.ArgumentParser:
 
 def check_header(filename: str, *, is_newly_created: bool = False) -> None:
   """Raises `HeaderCheckFailure` if the header doesn't match."""
+  expected_num_py2_lines = 6
+  expected_num_py3_lines = 3
   try:
     with open(filename, 'r') as f:
-      first_lines = [f.readline() for _ in range(0, 7)]
+      # We grab an extra line in case there is a shebang.
+      first_lines = [f.readline() for _ in range(0, expected_num_py2_lines + 1)]
   except IOError as e:
     raise HeaderCheckFailure(f"{filename}: error while reading input ({e})")
   # If a shebang line is included, remove it. Otherwise, we will have conservatively grabbed
   # one extra line at the end for the shebang case that is no longer necessary.
   first_lines.pop(0 if first_lines[0].startswith("#!") else - 1)
   # Check that the first lines even exists. Note that first_lines will always have an entry
   # for each line, even if the file is completely empty.
-  if len([line for line in first_lines if line]) < 4:
+  if len([line for line in first_lines if line]) < expected_num_py3_lines:
     raise HeaderCheckFailure(f"{filename}: missing the expected header")
+  is_py3_file = all("from __future__" not in line for line in first_lines)
+  if is_py3_file:
```

Comment:
I think I'd be more confident of the correctness of this script if it totally bifurcated at this point (i.e.
```
if is_py3_file:
  # end-to-end validation of Py3
else:
  # end-to-end validation of py23
```
MY worry is that one could craft a header such that it's not valid py2 or py3, but somehow passes `if "".join(first_lines) not in (EXPECTED_HEADER_PY2, EXPECTED_HEADER_PY3)`. Now, I realize this is impossible with the code in this change, just musing through a concern :)